pkg/controller/pv-monitor-controller/pv_monitor_controller.go

```diff
@@ -0,0 +1,240 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	handler "github.com/kubernetes-csi/external-health-monitor/pkg/csi-handler"
+	"github.com/kubernetes-csi/external-health-monitor/pkg/util"
+)
+
+// PVMonitorController is the struct of pv monitor controller containing all information to perform volumes health condition checking
+type PVMonitorController struct {
+	client             kubernetes.Interface
+	monitorName        string
+	monitorInterval    time.Duration
+	eventRecorder      record.EventRecorder
+	supportListVolumes bool
+
+	pvChecker *handler.PVHealthConditionChecker
+
+	nodeWatcher *NodeWatcher
+
+	csiConn *grpc.ClientConn
+
+	pvLister       corelisters.PersistentVolumeLister
+	pvListerSynced cache.InformerSynced
+
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	pvcListerSynced cache.InformerSynced
+
+	podLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	sync.Mutex
+	pvEnqueued map[string]bool
+
+	pvcToPodsCache *util.PVCToPodsCache
+
+	pvQueue workqueue.Interface
+}
+
+const (
+	listVolumesInterval     = 5 * time.Minute
+	pvReconcileSyncInterval = 10 * time.Minute
+)
+
+// NewPVMonitorController create PV monitor controller
+func NewPVMonitorController(client kubernetes.Interface, monitorName string, supportListVolumes bool, conn *grpc.ClientConn, timeout time.Duration, monitorInterval time.Duration, pvInformer coreinformers.PersistentVolumeInformer,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer, podInformer coreinformers.PodInformer, nodeInformer coreinformers.NodeInformer) *PVMonitorController {
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: client.CoreV1().Events(v1.NamespaceAll)})
+	var eventRecorder record.EventRecorder
+	eventRecorder = broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("csi-pv-monitor-controller-%s", monitorName)})
+
+	ctrl := &PVMonitorController{
+		csiConn:            conn,
+		eventRecorder:      eventRecorder,
+		supportListVolumes: supportListVolumes,
+		client:             client,
+		monitorName:        monitorName,
+		monitorInterval:    monitorInterval,
+		pvQueue:            workqueue.NewNamed("csi-monitor-pv-queue"),
+		//pvQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-monitor-pv-queue"),
+
+		pvcToPodsCache: util.NewPVCToPodsCache(),
+		pvEnqueued:     make(map[string]bool),
+	}
+
+	// PV informer
+	pvInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ctrl.pvAdded,
+		// UpdateFunc: ctrl.pvUpdated, TODO: do we need this?
```

Comment:
I think we need UpdateFunc as well. Otherwise we are assuming PVCs are only created after the health monitor is started.  We need to monitor volumes created before the health monitor is started as well.

I wonder if we should loop thru all bound PVCs in the API server and add them to the queue first when initializing the health monitor, and after that only wait for AddFunc to add additional ones?  If we rely only on AddFunc and UpdateFunc, we could miss volumes that are not newly created or being updated?

Reply:
We do not need to get all PVCs when initializing, informer(reflector) will help us do that.
Also, update is not needed in this case

Reply:
Please add what you have said here as comments and remove AddFunc and DeleteFunc